File: 0022 括号生成/main.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if(n==0):
            return []
        rl=[]
        def recursive(lparentheseLeft,rparentheseLeft,s):
            if(lparentheseLeft==0):
                rl.append(s+rparentheseLeft*")")
                return
            if(rparentheseLeft>lparentheseLeft):
                recursive(lparentheseLeft-1,rparentheseLeft,s+'(')
                recursive(lparentheseLeft,rparentheseLeft-1,s+')')
            else:
                recursive(lparentheseLeft-1,rparentheseLeft,s+'(')
        recursive(n,n,'')
        return rl

File: 0022 括号生成/main.cpp
class Solution {
    vector<string> res;
void genrecurision(string s,int lp,int rp){
    if(lp<=0 && rp<=0){
        res.push_back(s);
        return;
    }

    if(lp<rp ){
        if(lp>0)
            genrecurision(s+"(",lp-1,rp);
        genrecurision(s+")",lp,rp-1);
    }else{
        genrecurision(s+"(",lp-1,rp);
    }
}
public:
    vector<string> generateParenthesis(int n) {
        genrecurision("",n,n);
        return res;
    }
};